lib/stripe/api.ex

```diff
@@ -377,13 +377,21 @@ defmodule Stripe.API do
 
   defp do_perform_request_and_retry(method, url, headers, body, opts, {:attempts, attempts}) do
     response =
-      :telemetry.span(~w[stripe request]a, %{url: url, method: method}, fn ->
+      :telemetry.span(~w[stripe request]a, %{}, fn ->
+        telemetry_meta = %{
+          endpoint: URI.parse(url).path,
+          method: method,
+          attempt: attempts,
+          stripe_api_version: headers["Stripe-Version"],
+          status: nil
+        }
+
         case http_module().request(method, url, Map.to_list(headers), body, opts) do
```

Comment:
```suggestion
        case http_module().request(method, url, Map.to_list(headers), body, opts) do
```

Notice the usage of `http_module()` which means this level of abstraction should be only about Stripe specific info (such as the header?!) but if people use `Req` or other clients, such `http_module()` may include its own Telemetry component.

Reply:
Telemetry should work fine with other clients, it's opt in and people need to configure their own handlers in code to use these events so there is no risk of conflict with other libs having their own telemetry. Metrics will ofc overlap to some extent, but our events have additional data that makes them meaningful in Stripe sense, such as API endpoint and version or event kind for webhooks. 

I added Stripe version header because it can uniquely identify API endpoint, dunno if anyone would miss it if it's not here. Actually status and request path are most useful, I can remove all others if you like.

You could get exactly same data by creating your own http_module and webhook handler and adding telemetry in there, but having it available out of the box and without need for custom adapters or other code is nice to have and has value imo.

Reply:
My point is that we could add the telemetry and lower its location, closer to the HTTP client itself; in the case of `hackney,` we need to provide the telemetry implementation; other clients like `Req` could use whatever available built-ins or add the layer to it.

What that means in practice is to change the following function to use our hackney module:

```diff
  @spec http_module() :: module
  def http_module() do
-    Config.resolve(:http_module, :hackney)
+    Config.resolve(:http_module, Stripe.Hackney) # <----- add the callback and telemetry
  end
```

I found a package that adds `:telemetry` to hackney already: https://github.com/msramos/hackney_telemetry

Reply:
Hmm, so the way it is now you can say that it adds layer to whatever http client provides. Users will be able to use our or their http client telemetry or both at the same time, we don't need to provide per client implementation.

If we don't add telemetry then users need to filter generic events for stripe requests probably based on path or hostname if these attributes are available. I bet many will be miserable doing that 🙃 😆 

If we add our telemetry then they can use our events and don't need to think about filtering and what events look like in their http client library, it will take just few lines of code to build dashboards like I shown in PR description. I could also can add a section to readme make our telemetry easier to use.

I think pushing code one level down makes sense if the goal is to clean up API module and push there more http concerns, for example retries, but functionally telemetry would continue to work the same as now, that is will be available regardless of telemetry provided by actual http client.